### PR TITLE
[todo] add rand(), srand()

### PR DESCRIPTION
**Pull request by mattn**
This add [Xorshift](https://en.wikipedia.org/wiki/Xorshift) random generator. In general implementation, srand() may be called from some plugins in several times. So it need to have statefull random generator like Xorshift. Below is an usage of srand().

```vim
:echo srand()
[123456789, 362436069, 521288629, 88675123]
:echo srand(-1)
[1479837050,, 362436069, 521288629, 88675123]
:echo srand(2)
[2, 362436069, 521288629, 88675123]
```
srand return four numbers as list. When expr is not given, the x element of xorshift become 123456789. When negative number is given, x become time(NULL). When positive number is given, x become it given.

Below is an usage of rand()

```vim
let a = srand()
:echo rand(a)
597902826
:echo rand(a)
458295558
:echo rand(a)
1779455562
:echo rand(a)
663552176
:echo rand(a)
507026878
```

### CODE REVIEW

**Comment by brammool** [comment]
Yasuhiro Matsumoto wrote:

> This add Xorshift random generator. In general implementation, srand()
> may be called from some plugins in several times. So it need to have
> statefull random generator like Xorshift. Below is an usage of
> srand().
> 
> ```vim
> :echo srand()
> [123456789, 362436069, 521288629, 88675123]
> :echo srand(-1)
> [1479837050,, 362436069, 521288629, 88675123]
> :echo srand(2)
> [2, 362436069, 521288629, 88675123]
> ```
> srand return four numbers as list. When expr is not given, the x element of xorshift become 123456789. When negative number is given, x become time(NULL). When positive number is given, x become it given.
> 
> Below is an usage of rand()
> 
> ```vim
> let a = srand()
> :echo rand(a)
> 597902826
> :echo rand(a)
> 458295558
> :echo rand(a)
> 1779455562
> :echo rand(a)
> 663552176
> :echo rand(a)
> 507026878
> ```

What plugin needs a pseudo-random number generator?

-- 
hundred-and-one symptoms of being an internet addict:
43. You tell the kids they can't use the computer because "Daddy's got work to
    do" and you don't even have a job.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
For example,

## lorem ipsum generator

https://github.com/mattn/emmet-vim

This generate example strings. It's useful to confirm html/css with texts like below.

>Ipsum in quidem aliquid quas obcaecati. Necessitatibus ullam excepturi deserunt numquam quod. Eligendi aperiam repellendus deserunt aliquam voluptate possimus ut quibusdam veniam, magni Exercitationem nesciunt hic esse asperiores suscipit Quidem!

## random tips display

https://github.com/vim-scripts/Random-Tip-Displayer

This is useful to display random tips when vim startup.

## random colorscheme

https://github.com/mattn/random_colorscheme-vim

This may be useful to change of pace, maybe. :smile: 

## games

fun with vim. :video_game: 

invador https://github.com/mattn/invader-vim
samegame https://github.com/ynkdir/vim-samegame
mahjong https://github.com/mattn/mahjong-vim
tetris http://www.vim.org/scripts/script.php?script_id=172
sokoban http://www.vim.org/scripts/script.php?script_id=211
mines http://www.vim.org/scripts/script.php?script_id=551
lifegame http://www.vim.org/scripts/script.php?script_id=377
sudoku http://www.vim.org/scripts/script.php?script_id=3553
https://github.com/mattn/flappyvird-vim

**Comment by mattn** [comment]
In other case, useful to generate test data.

**Comment by kamichidu** [comment]
Somtimes I want to generate UUID for editing some files, html element id, visual studio solution file for example.
If I have these functions, I can create UUIDv4 implementation.

I know you probably say I can implement UUIDv1,2,3 and 5.
But I think UUIDv4 is better than others, since it reduce conflicts under actual stuation.

File: src/evalfunc.c [hunk 1]
```diff
@@ -8046,6 +8052,52 @@ f_pyeval(typval_T *argvars, typval_T *rettv)
...
 
 /*
+ * "rand()" function
+ */
+    static void
+f_rand(typval_T *argvars, typval_T *rettv)
+{
+    list_T	*l;
+    listitem_T  *lx, *ly, *lz, *lw;
+    long	x, y, z, w, t;
```
Comment (ichizok) [inline src/evalfunc.c]: These must be (32bit) unsigned int.

**Comment by mattn** [comment]
I don't think the naming rand/srand is good. If you have an idea, please let me know.

**Comment by chrisbra** [comment]
+1 for a random generator. Considering there have been patches posted in the past already several times I think it is generally useful

File: src/evalfunc.c [hunk 2]
```diff
@@ -11037,8 +11119,37 @@ f_sqrt(typval_T *argvars, typval_T *rettv)
...
+ * "rand()" function
+ */
+    static void
+f_srand(typval_T *argvars, typval_T *rettv)
+{
+    if (rettv_list_alloc(rettv) == FAIL)
+	return;
+    if (argvars[0].v_type == VAR_UNKNOWN)
+	list_append_number(rettv->vval.v_list, 123456789);
+    else if (argvars[0].v_type == VAR_NUMBER)
```
Comment (ZyX-I) [inline src/evalfunc.c]: This should use `get_tv_number_chk`, for Vim functions it is expected to automatically convert strings to numbers.

File: src/evalfunc.c [hunk 3]
```diff
@@ -11037,8 +11119,37 @@ f_sqrt(typval_T *argvars, typval_T *rettv)
...
+    {
+	long x = (long)argvars[0].vval.v_number;
+	if (x < 0)
+	    list_append_number(rettv->vval.v_list, (varnumber_T)time(NULL));
+	else
+	    list_append_number(rettv->vval.v_list, argvars[0].vval.v_number);
+    }
+    else {
```
Comment (ZyX-I) [inline src/evalfunc.c]: Currently used style requires putting `{` on the dedicated line.

**Comment by ZyX-I** [comment]
I also needed random number generator a few times, mostly imitating it with `reltimestr()` AFAIR.

**Comment by mattn** [comment]
Xorshift use full 32bit and Vim script deson't have unsigned number. So this seems not work on `-num64`.

**Comment by mattn** [comment]
Do you prefer xorshift() than rand() ?

**Comment by ichizok** [comment]
I could be wrong, but;
what matters are, bit-pattern to be random, and calculating by logical-shift in c-code,
so Vim script perhaps can use random as singed even when without num64.

File: src/evalfunc.c [hunk 4]
```diff
@@ -11041,8 +11120,37 @@ f_sqrt(typval_T *argvars, typval_T *rettv)
...
+f_srand(typval_T *argvars, typval_T *rettv)
+{
+    if (rettv_list_alloc(rettv) == FAIL)
+	return;
+    if (argvars[0].v_type == VAR_UNKNOWN)
+	list_append_number(rettv->vval.v_list, 123456789);
+    else
+    {
+	int error = FALSE;
+	UINT32_TYPEDEF  x = (UINT32_TYPEDEF)get_tv_number_chk(&argvars[0], &error);
```
Comment (brcolow) [inline src/evalfunc.c]: Did you mean to have two spaces after `UINT32_TYPEDEF`?

File: src/evalfunc.c [hunk 5]
```diff
@@ -8050,6 +8056,79 @@ f_pyeval(typval_T *argvars, typval_T *rettv)
 #endif
 
 /*
+ * "rand()" function
+ */
+    static void
+f_rand(typval_T *argvars, typval_T *rettv)
+{
+    list_T	*l = NULL;
+    UINT32_TYPEDEF  x, y, z, w, t;
```
Comment (brcolow) [inline src/evalfunc.c]: Same here.

File: src/evalfunc.c [hunk 6]
```diff
@@ -8050,6 +8056,79 @@ f_pyeval(typval_T *argvars, typval_T *rettv)
...
+
+#define SHUFFLE_XORSHIFT128 \
+	t = x ^ (x << 11); \
+	x = y; y = z; z = w; \
+	w = (w ^ (w >> 19)) ^ (t ^ (t >> 8));
+
+    if (argvars[0].v_type == VAR_UNKNOWN)
+    {
+	/* When argument is not given, return random number initialized
+	 * staticly. */
```
Comment (brcolow) [inline src/evalfunc.c]: staticly => statically

File: src/evalfunc.c (hunk 4 above)
Comment (mattn) [reply src/evalfunc.c]: done

File: src/evalfunc.c (hunk 5 above)
Comment (mattn) [reply src/evalfunc.c]: done

File: src/evalfunc.c (hunk 6 above)
Comment (mattn) [reply src/evalfunc.c]: done

**Comment by chrisbra** [comment]
This pull request **introduces 1 alert** when merging b5aa623ab12eabd5a0187725245e3122ef8f05fb into 895d966e341c187ffcf9da88dba193cbfcebf000 - [view on LGTM.com](https://lgtm.com/projects/g/vim/vim/rev/pr-784ed9bc218776a044fbd53f77c2836999da93be)

**new alerts:**

* 1 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by chrisbra** [comment]
This pull request **introduces 1 alert** when merging 2520cda68c22d1e6da46af2ef9cf1e6750d59e49 into b0e2da2b23e7497d0a145ecce9c1e08b6fab6dbf - [view on LGTM.com](https://lgtm.com/projects/g/vim/vim/rev/pr-f892323b6bd94903ef32db661c49ce7b7a6a6b7a)

**new alerts:**

* 1 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/evalfunc.c [hunk 7]
```diff
@@ -12524,7 +12603,33 @@ f_sqrt(typval_T *argvars, typval_T *rettv)
     else
 	rettv->vval.v_float = 0.0;
 }
+#endif
+
+/*
+ * "rand()" function
```
Comment (k-takata) [inline src/evalfunc.c]: s/rand/srand/

**Comment by brammool** [comment]
It's been a long time in the todo list, and finally bubbled up.
Instead of UINT32_TYPEDEF it should use UINT32_T.
I suppose it's OK to restrict to 32 bit numbers, to that the code using it is portable.
There could be a few more test cases, e.g. with an invalid argument to rand().

To be a bit more descriptive it should be called pseudo-random.
Can someone comment on the algorithm used?

**Comment by mattn** [comment]
Fixed implementation to be possible to use easily.

* srand() without {expr} initialize global seed value
* rand() without {expr} generate pseudo random value with global seed value
* srand() with {expr} initialize seed value with x element by given value
* rand() with {expr} generate pseudo random value with seed value given

>Can someone comment on the algorithm used?

https://en.wikipedia.org/wiki/Xorshift

**Comment by umireon** [comment]
xoshiro128** is better than the Marseglia's xorshift in the point of speed and equidistribution.
For detail, see http://prng.di.unimi.it/

**Comment by mattn** [comment]
@brammool Which do you prefer xorshift or xoshiro128**

**Comment by umireon** [comment]
Initializing the state with a fixed number is not good because the similar bit pattern produces the similar random sequences.
For example, srand(1) and srand(2) produce similar sequences in spite of expectation.

The most famous state initialization algorithm might be [MT's init_genrand](http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/MT2002/CODES/mt19937ar.c), but I suggest that using [SplitMix32](https://github.com/umireon/my-random-stuff/blob/master/xorshift/splitmix32.c)  to fill the state.

The example of the proper state initialization is placed in [vital.vim](https://github.com/vim-jp/vital.vim/blob/master/autoload/vital/__vital__/Random/Xor128.vim#L46).

**Comment by brammool** [comment]
> Initializing the state with a fixed number is not good because the
> similar bit pattern produces the similar random sequences.
> For example, srand(1) and srand(2) produce similar sequences in spite
> of expectation.
> 
> The most famous state initialization algorithm might be [MT's init_genrand](http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/MT2002/CODES/mt19937ar.c), but I suggest that using [SplitMix32](https://github.com/umireon/my-random-stuff/blob/master/xorshift/splitmix32.c)  to fill the state.
> 
> The example of the proper state initialization is placed in [vital.vim](https://github.com/vim-jp/vital.vim/blob/master/autoload/vital/__vital__/Random/Xor128.vim#L46).

Can you suggest a patch?

-- 
Would you care for a drink?   I mean, if it were, like,
disabled and you had to look after it?

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by umireon** [comment]
OK, I will.

**Pull request closed without merging**